src/fft.rs

```diff
@@ -0,0 +1,200 @@
+use crate::util::parallelize;
+use ff::{Field, PrimeField};
+use group::{GroupOpsOwned, ScalarMulOwned};
+pub use halo2curves::{CurveAffine, CurveExt};
+
+/// Given FFT domain size k, return the omega in case of fft
+/// or return the omega_inv in case if ifft
+pub(crate) fn get_omega_or_inv<F: PrimeField>(k: u32, is_inverse: bool) -> F {
+    let mut omega_or_inv = if is_inverse {
+        F::ROOT_OF_UNITY_INV
+    } else {
+        F::ROOT_OF_UNITY
+    };
+    for _ in k..F::S {
+        omega_or_inv = omega_or_inv.square();
+    }
+    omega_or_inv
+}
+
+pub(crate) fn get_ifft_divisor<F: PrimeField>(k: u32) -> F {
+    F::TWO_INV.pow_vartime([k as u64, 0, 0, 0])
+}
+
+/// This represents an element of a group with basic operations that can be
+/// performed. This allows an FFT implementation (for example) to operate
+/// generically over either a field or elliptic curve group.
+pub trait FftGroup<Scalar: Field>:
+    Copy + Send + Sync + 'static + GroupOpsOwned + ScalarMulOwned<Scalar>
+{
+}
+
+impl<T, Scalar> FftGroup<Scalar> for T
+where
+    Scalar: Field,
+    T: Copy + Send + Sync + 'static + GroupOpsOwned + ScalarMulOwned<Scalar>,
+{
+}
+
+/// Performs a radix-$2$ Fast-Fourier Transformation (FFT) on a vector of size
+/// $n = 2^k$, when provided `log_n` = $k$ and an element of multiplicative
+/// order $n$ called `omega` ($\omega$). The result is that the vector `a`, when
+/// interpreted as the coefficients of a polynomial of degree $n - 1$, is
+/// transformed into the evaluations of this polynomial at each of the $n$
+/// distinct powers of $\omega$. This transformation is invertible by providing
+/// $\omega^{-1}$ in place of $\omega$ and dividing each resulting field element
+/// by $n$.
+///
+/// This will use multithreading if beneficial.
+pub(crate) fn best_fft<Scalar: Field, G: FftGroup<Scalar>>(a: &mut [G], omega: Scalar, log_n: u32) {
```

Comment:
fft implementation looks correct, but I don't think radix 2 is the most performant

Reply:
This code is from halo2curves. In our case, the NTT size will not be very large comparing the halo2_proofs. We might consider using a better version if find it needed later.

Reply:
"if there is a case when you need different size of ntt, would not it be better to keep nth root of unities?

I don't get this part. Could you explain more on it? @bbarbakadze

Reply:
I mean you could hardcode or precalculate and store `domain_size_inverse` values `forward_roots_of_unity` and `inverse_roots_of_unity` in case you need to calculate them several times in your code.

Reply:
Ah, I see. Let's see how many times needed. I added a TODO if we need multiple times to call